### PR TITLE
CNF-12206: introduce new .status.history field to understand timing of each stage and their phases

### PR DESCRIPTION
**Pull request by pixelsoccupied**
# Background / Context

A new status `history` field is added to understand when and for how long each stage may run successfully.  

# Solution / Feature Overview

- History is reset when `Idle`.
- Each history entry on the list will appear chronologically and will have `startTime` to indicated when a stage was called for the first time. Optionally `completionTime` will appear when the stage is successful.
- A stage may have a one or more `Phase` to help understand important tasks that were performed. Similar to history entries, each `Phase` will have a `startTime` and optionally a `completionTime`.
    - 'Prep' stage entry will have the follow:
        - `Stateroot`: Time taken to create the new stateroot.
        - `Precache`: Time taken to pull all images specified by the image list built into the seed image.
    - 'Upgrade' stage entry will have the follow:
        - `PrePivot`: Time taken complete all the steps before reboot is called
        - `PostPivot`: Time taken complete all the steps after a reboot to a new stateroot
        - Reboot time can be calculated with PrePivot completionTime and PostPivot startTime.

# Other Information

/cc @Missxiaoguo @jc-rh @donpenney

### CODE REVIEW

File: docs/image-based-upgrade.md [hunk 1]
```diff
@@ -211,6 +211,49 @@ status:
   - Prep
 ```
 
+IBU status `.histoy` can be used understand when stages were desired and time taken to complete successfully
```
Comment (donpenney) [inline docs/image-based-upgrade.md:214]: Typo
```suggestion
IBU status `.history` provides timing information for the stages, indicating start and completion times for the stages
```

File: docs/image-based-upgrade.md [hunk 2]
```diff
@@ -211,6 +211,49 @@ status:
...
+    - completionTime: "2024-06-26T03:48:12Z"
+      stage: Rollback
+      startTime: "2024-06-26T03:40:42Z"
+```
+
+Things to note from about `.status.history`
+
+- History is reset when desired stage is`Idle`.
+- Each history entry on the list will appear chronologically and will have `startTime` to indicated when a stage was desired for the first time. Optionally `completionTime` will appear when a stage completes successfully.
+- A stage may have a one or more `Phase` to help understand important tasks that were performed. Similar to history entries, each `Phase` will have a `startTime` and optionally a `completionTime` when completes successfully.
```
Comment (donpenney) [inline docs/image-based-upgrade.md:248]:
```suggestion
- A stage may have one or more `Phases` to help understand important tasks that were performed. Similar to history entries, each `Phase` will have a `startTime` and, once completed, a `completionTime`.
```

File: docs/image-based-upgrade.md [hunk 3]
```diff
@@ -211,6 +211,49 @@ status:
...
+      startTime: "2024-06-26T03:27:36Z"
+    - completionTime: "2024-06-26T03:48:12Z"
+      stage: Rollback
+      startTime: "2024-06-26T03:40:42Z"
+```
+
+Things to note from about `.status.history`
+
+- History is reset when desired stage is`Idle`.
+- Each history entry on the list will appear chronologically and will have `startTime` to indicated when a stage was desired for the first time. Optionally `completionTime` will appear when a stage completes successfully.
```
Comment (donpenney) [inline docs/image-based-upgrade.md:247]:
```suggestion
- Each history entry on the list will appear chronologically and will have `startTime` to indicate when a stage was triggered. The `completionTime` will appear once a stage completes successfully.
```

Comment (donpenney) [inline docs/image-based-upgrade.md:247]: What about when the stage fails?

Comment (pixelsoccupied) [reply docs/image-based-upgrade.md]: So if things fail we will only have startTime (and just reset after moving to `Idle`). 

It's similar to Job's status where [completionTime](https://github.com/kubernetes/api/blob/master/batch/v1/types.go#L526) only appears when it completes successfully.

Comment (pixelsoccupied) [reply docs/image-based-upgrade.md]: Yeah! Basically the way i was reading it...."stop the timer" with stopABC to complete. Do you want me to rename?

Comment (jc-rh) [reply docs/image-based-upgrade.md]: don't bother if no other changes required

**Review by donpenney — APPROVED** [review]

**Pull request merged**